### ISSUE TITLE
[BUG] Blazor ReactiveComponentBase throws errors

### ISSUE THREAD

**Issue by richbryant** [reactiveui/ReactiveUI#2308]
**Describe the bug**
In ReactiveUI.Blazor, ReactiveComponentBase now throws errors on both server and client side.

This is new behaviour.  It didn't before.  You can bypass by implementing IViewFor in code-behind but the component itself ist kaput.

I'll investigate further and add details here as I find them.

**Comment by glennawatson** [comment]
Nice header would be nice :)

**Comment by richbryant** [comment]
Sorry.  Note to self, don't file bugs at 1:30am.

**Comment by richbryant** [comment]
alright, not only does that could break on StartUp - due to injection of VM being slower than firing events in the constructor - but, and this is important **it shouldn't be there anyway**.

If you check out the projects I have in [Samples](https://github.com/reactiveui/ReactiveUI.Samples/tree/master/blazor), these all implement `IViewFor<TViewModel>` rather than inheriting ReactiveComponentBase, none of them have any code issuing a StateHasChanged() and this is the correct behaviour.  They work.

To explain, StateHasChanged() is _global state_.  If you invoke it, the whole page re-renders.  This is not something we want to be happening when, for example, only the incremented Counter figure has changed.  More pertinently, it's very common to nest components inside components and if you want the parent component to re-render, you should explicitly say so (ie, fire StateHasChanged() yourself) - full page re-render should not be the default behaviour since on wasm it's a considerable and unnecessary performance hit and on ServerSide, it's a huge increase in the size of the delivered delta _and_ a performance hit.

I recommend we remove the StateHasChanged section from ReactiveComponentBase entirely.

**Comment by richbryant** [comment]
Now that I think about it, you could easily just call `StateHasChanged` on the `ViewModel` property's setter.  That solves @limefrogyank 's issue.

**Comment by limefrogyank** [comment]
I haven't updated to this version yet because I've just been using my own version of `ReactiveComponentBase`.  It's actually very similar.

The only difference is the order in which I subscribe to the ViewModel property change and that I check for `WhereNotNull()` on it before calling subscribe.

```
public ReactiveComponentBase()
{
     var propertyChangedObservable = this.WhenAnyValue(x => x.ViewModel)
            .WhereNotNull()
            .Select(x => Observable.FromEvent<PropertyChangedEventHandler, Unit>(
                eventHandler =>
                {
                    void Handler(object sender, PropertyChangedEventArgs e) => eventHandler(Unit.Default);

                    return Handler;
                },
                eh => x.PropertyChanged += eh,
                eh => x.PropertyChanged -= eh))
            .Switch();

        propertyChangedObservable.Do(_ => StateHasChanged()).Subscribe();

        this.WhenAnyValue(x => x.ViewModel).WhereNotNull().Subscribe(_ => StateHasChanged());
}
```
I don't think there's any way to get around calling `StateHasChanged()` in either section.  That's the only way to update the UI.  While there's overhead in checking the differences, the renderer will only update the parts of the UI that have actually changed.

I can't pass in the ViewModel in the page constructor, so I need a `StateHasChanged()` to let the already rendered page know the ViewModel property is no longer null.  In the `SextantRouter` I'm working on, the router creates `ReactiveRouteView`s which host the actual page that is being rendered and the ViewModel is set using a callback from the `AddComponentReferenceCapture` method on the `RenderTreeBuilder` object:
https://github.com/limefrogyank/Sextant/blob/Blazor/src/Sextant.Blazor/ReactiveRouteView.cs#L96-L134

In any case, this has been working pretty well for me.

**Comment by richbryant** [comment]
if it's working, it's working!

This looks like it's yours to PR.  I'd certainly like to swap out the IViewFor in the samples.

**Comment by limefrogyank** [comment]
Also, I haven't really come up with a good way of avoiding all the null checks for `ViewModel` in the razor code itself.

I use `ViewModel?.someProperty` for most things, but for `@bind-Value=ViewModel.someProperty`, you have to explicity check for null with a `@if (ViewModel != null)`.  I can't get the null-check operator to work on those.

**Comment by richbryant** [comment]
btw, there is an issue raised to allow constructor injection on Components now that we have `partial` support.

**Comment by limefrogyank** [comment]
> > the renderer will only update the parts of the UI that have actually changed.
> 
> Sadly not true - if there's a parent component, that will also get rerendered.

But lifecycle events are not called.  My `SextantRouter` which is the parent of all components only runs `OnAfterRenderAsync` once really.  Anytime it renders a new page, `OnAfterRenderAsync` isn't being triggered at all.

Beyond that, I'm still not sure how the underlying renderer works...

**Comment by richbryant** [comment]
Interesting.  Ok, in the meantime I'll fork Reactive.Blazor, try integrating your example and see what the results are.  Unless you want to?

**Comment by limefrogyank** [comment]
No, no... you do it, please.  I'm terrible at Git commands.

**Comment by richbryant** [comment]
Done.

**Comment by limefrogyank** [comment]
I had assumed it was part of System.Reactive.Linq now I see it was part of the Sextant library.  

```
public static IObservable<T> WhereNotNull<T>(this IObservable<T> observable)
        {
            return observable.Where(x => x != null);
        }
```

**Comment by richbryant** [comment]
Welp.  That didn't work.  I had it running but from the Nuget, it's a bust.  I'll see if I can find out what's going on.  Stick to your version for now, @limefrogyank .

**Comment by glennawatson** [comment]
I haven't pushed any releases yet.

**Comment by richbryant** [comment]
that's weird, I wonder the nuget wanted to update?

Must have been on an old one.  Okay, ignore me.

**Comment by limefrogyank** [comment]
Hate to be the party pooper, but I finally updated to the latest ReactiveUI.Blazor package.  I know the latest nuget doesn't have the fix yet, so I am still using my own version of `ReactiveComponentBase`.  In wasm, it's fine (I think).  However, in server-side blazor, I'm getting an exception which needs to be solved using an `InvokeAsync(StateHasChanged)` instead of pure `StateHasChanged()` for the ViewModel property changing.

```
this.WhenAnyValue(x => x.ViewModel).WhereNotNull().Subscribe(_ => InvokeAsync(StateHasChanged));
```

**Comment by glennawatson** [comment]
What exception are you getting?

I suspect this WhenAnyValue() may need to be moved to the OnInitialize() if it's the render hook bug.

**Comment by limefrogyank** [comment]
It happens when my view has already been created and the router is manually setting the ViewModel property with the correct viewmodel.  A dispatcher error:
```
Error: System.InvalidOperationException: The current thread is not associated with the Dispatcher. Use InvokeAsync() to switch execution to the Dispatcher when triggering rendering or component state.
```

**Comment by limefrogyank** [comment]
Or instead of using `InvokeAsync`, this works, too.
```
this.WhenAnyValue(x => x.ViewModel).WhereNotNull().ObserveOn(SynchronizationContext.Current).Subscribe(_ => StateHasChanged());
```

**Comment by limefrogyank** [comment]
Scratch that.  It kills the error, but it doesn't always work to update the UI.  Only InvokeAsync(StateHasChanged) has worked consistently for me

**Comment by richbryant** [comment]
Considering adding a StateChanged<Unit, Unit> ReactiveCommand in ReactiveComponentBase which is simply `InvokeAsync(StateHasChanged)` for clarity.

This would also allow "manual" invocations if it's public and be a handy hook if it's `virtual`.

**Comment by glennawatson** [comment]
Closing the issue since it should be fixed by RxUI 11.1.23 -- comment back if it's not.

**Comment by richbryant** [comment]
It's not, dammit.

This "swap out ViewModels on the fly" things has crippled it.
